zaptest/tlogger/tlogger.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package tlogger provides a zapcore.Core that is capable of writing log
+// messages to a *testing.T and *testing.B. It may be used from Go tests or
+// benchmarks to have log messages printed only if a test failed, or if the
+// `-v` flag was passed to `go test`.
+package tlogger // import "go.uber.org/zap/zaptest/tlogger"
+
+import (
+	"testing"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// New builds a new Core that logs all messages to the given testing.TB.
+//
+// Use this with a *testing.T or *testing.B to get logs which get printed only
+// if a test fails or if you ran go test -v.
+//
+//   logger := zap.New(tlogger.New(t))
+func New(t testing.TB) zapcore.Core {
```

Comment:
could this just take `...zap.Option`:
https://godoc.org/go.uber.org/zap#Option

that way, users could pass a level since it implements option, but they can also pass other options to add callers, stacktraces, etc.

Reply:
Couple reasons it's like this:

- zap.Option doesn't let you override the level.
- [`zaptest/observer.New`](https://godoc.org/go.uber.org/zap/zaptest/observer#New) returns a zapcore.Core too.
- Instead of `New` always accepting a level, I wanted a "just works" version which is why we have both `New` and `NewAt`.

Reply:
Fair enough. I wonder if a `zap.Logger` is more useful than a core most of the time (and you can still get the core using `logger.Core`
https://godoc.org/go.uber.org/zap#Logger.Core

I thought levels were options, but if not, `NewAt` definitely makes sense.

Reply:
I don't feel strongly about `Logger` vs `Core`. @akshayjshah what do you think?

Reply:
I don't feel super-strongly about it either; I suggested returning a `Core` here to match the `observer.New` API.

Reply:
I think we should deprecate the `observer` package, and as part of https://github.com/uber-go/zap/issues/520, introduce a set of options that can be used with either the verifying logger, or to filter a set of observed logs.

As part of this, we can introduce a more ergonomic API which is probably to return a `Logger` (every test here, and 99% of use cases where we use the observer want to use it as a `*Logger`). Since it's pretty easy to get a core out of a logger, it doesn't limit functionality.